src/main/kotlin/org/jetbrains/kotlinx/jupyter/repl/impl/InternalEvaluatorImpl.kt

```diff
@@ -124,4 +158,73 @@ internal class InternalEvaluatorImpl(
             isExecuting = false
         }
     }
+
+    private fun updateVariablesState(cellId: Int) {
+        // remove known modifying usages in this cell
+        usageMap[cellId]?.removeIf {
+            containingMap[it] != cellId
+        }
+
+        variablesMap.forEach {
+            val state = it.value
+            val property = state.propertyKlass
+            val newValue = property.get(state.scriptInstance) ?: return@forEach
+
+            if (state.stringValue != newValue.toString()) {
+                usageMap[cellId]?.add(it.key)
+                state.stringValue = newValue.toString()
+            }
+        }
+    }
+
+    private fun getVisibleVariables(target: ResultValue, cellId: Int): Map<String, VariableState> {
+        val klass = target.scriptClass ?: return emptyMap()
+        val cellKlassInstance = target.scriptInstance!!
+
+        val fields = klass.declaredMemberProperties
+        val ans = mutableMapOf<String, VariableState>()
+        fields.forEach {
+            val casted = it as KProperty1<Any, *>
+            ans[casted.name] = if (casted.get(cellKlassInstance) == null) {
+                return@forEach
```

Comment:
What about changing `.forEach`-`return@forEach` to simple `for`-`continue`?

Reply:
Let's discuss! From what I know about Java, all these lambdas are stored in an inner class inside a calling class. Yes, it may sound like a solution with overhead to resolve needed one and call a method, but I rely on the fact we do it repeatedly. 
So, as it seems to me, both these solutions are equal, being the only difference that lambda with that `return-to-label` being a bit fancier

Reply:
I'm not talking about performance) If you really love this `return@forEach` construction, leave it as is, but I'd use for-continue here